Framework/Core/include/Framework/DataAllocator.h

```diff
@@ -217,171 +217,109 @@ class DataAllocator
+        auto sizeInBytes = elementSizeInBytes * object.size();
+        FairMQMessagePtr payloadMessage(proxy.createMessage(sizeInBytes));
+
+        if constexpr (std::is_pointer<typename T::value_type>::value == false) {
+          // vector of elements
+          memcpy(payloadMessage->GetData(), object.data(), sizeInBytes);
+        } else {
+          // serialize vector of pointers to elements
+          auto target = reinterpret_cast<unsigned char*>(payloadMessage->GetData());
+          for (auto const& pointer : object) {
+            memcpy(target, pointer, elementSizeInBytes);
+            target += elementSizeInBytes;
+          }
+        }
 
-  /// Explicitely ROOT serialize a snapshot of @a object when called,
-  /// will then be sent once the computation ends. The @a object is wrapped
-  /// into type ROOTSerialized to explicitely mark this serialization method,
-  /// and is expected to have a ROOT dictionary. Availability can not be checked
-  /// at compile time for all cases.
-  /// Framework does not take ownership of the @a object. Changes to @a object
-  /// after the call will not be sent.
-  template <typename W>
-  typename std::enable_if<is_specialization<W, ROOTSerialized>::value == true, void>::type
-    snapshot(const Output& spec, W wrapper)
-  {
-    using T = typename W::wrapped_type;
-    static_assert(std::is_same<typename W::hint_type, const char>::value || //
-                    std::is_same<typename W::hint_type, TClass>::value ||   //
-                    std::is_void<typename W::hint_type>::value,             //
-                  "class hint must be of type TClass or const char");
-
-    auto proxy = mContextRegistry->get<RootObjectContext>()->proxy();
-    FairMQMessagePtr payloadMessage(proxy.createMessage());
-    const TClass* cl = nullptr;
-    if (wrapper.getHint() == nullptr) {
-      // get TClass info by wrapped type
-      cl = TClass::GetClass(typeid(T));
-    } else if (std::is_same<typename W::hint_type, TClass>::value) {
-      // the class info has been passed directly
-      cl = reinterpret_cast<const TClass*>(wrapper.getHint());
-    } else if (std::is_same<typename W::hint_type, const char>::value) {
-      // get TClass info by optional name
-      cl = TClass::GetClass(reinterpret_cast<const char*>(wrapper.getHint()));
-    }
-    if (has_root_dictionary<T>::value == false && cl == nullptr) {
-      std::string msg("ROOT serialization not supported, dictionary not found for type ");
-      if (std::is_same<typename W::hint_type, const char>::value) {
-        msg += reinterpret_cast<const char*>(wrapper.getHint());
+        addPartToContext(std::move(payloadMessage), spec, o2::header::gSerializationMethodNone);
+      } else if constexpr (has_root_dictionary<ElementType>::value) {
+        return snapshot(spec, ROOTSerialized<T const>(object));
       } else {
-        msg += typeid(T).name();
+        static_assert(type_dependent<T>::value,
```

Comment:
Notice that I've renamed `type_dependent<T>` to `always_static_assert<T>` and moved it to `Framework/Traits.h`.

Reply:
ah, I have noticed it in the code and did a quick search without much effort, thanks!